juju1/state/migration_export.go

```diff
@@ -931,6 +944,76 @@ func (e *exporter) addApplication(ctx addApplicationContext) error {
 	return nil
 }
 
+func extraBindings(application *Service) ([]string, error) {
+	// Collect any extra bindings from the charm. These aren't
+	// captured in the db but they're stored in the charm blob, in the
+	// metadata file.
+	ch, _, err := application.Charm()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	st := application.st
+	store := statestorage.NewStorage(st.EnvironUUID(), st.MongoSession())
+	reader, _, err := store.Get(ch.StoragePath())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer reader.Close()
+
+	localFile, err := ioutil.TempFile("", "charm-"+ch.URL().Name)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer os.Remove(localFile.Name())
```

Comment:
just a single defer func with both cleanups in the one func?

Reply:
Yeah, will do - especially given that I had to check to remember which order the cleanups happen in.